Error while checking build/O2/fullCI for 8222c00d531b3b169aea53225e397cc735660086 at 2024-12-14 13:19:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2Physics-latest/log
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:


## sw/BUILD/O2-sim-challenge-test-latest/log
./sim-challenge.logDetected critical problem in logfile mftreco.log
./sim-challenge.logmftreco.log:[14846:internal-dpl-ccdb-backend]: [12:19:04][ERROR] Exception while running: Fatal error. Rethrowing.
./sim-challenge.logmftreco.log-[14846:internal-dpl-ccdb-backend]: [12:19:04][FATAL] Unhandled o2::framework::runtime_error reached the top of main of o2-mft-reco-workflow, device shutting down. Reason: Fatal error
./sim-challenge.log[14846:internal-dpl-ccdb-backend]: [12:19:04][ERROR] CCDBDownloader CURL transfer error - Timeout was reached
./sim-challenge.log[14846:internal-dpl-ccdb-backend]: [12:19:04][ERROR] CcdbDownloader finished transfer http://alice-ccdb.cern.ch/CTP/Calib/OrbitReset for 1546300800000 (agent_id: alissandra05-1734178742-hDz9SB) with http code: 0
./sim-challenge.log[14846:internal-dpl-ccdb-backend]: [12:19:04][ERROR] File CTP/Calib/OrbitReset could not be retrieved. No more hosts to try.
./sim-challenge.log[14846:internal-dpl-ccdb-backend]: [12:19:04][FATAL] Unable to find object CTP/Calib/OrbitReset/1546300800000
./sim-challenge.log[14846:internal-dpl-ccdb-backend]: [12:19:04][ERROR] Exception while running: Fatal error. Rethrowing.
./sim-challenge.log[14846:internal-dpl-ccdb-backend]: [12:19:04][FATAL] Unhandled o2::framework::runtime_error reached the top of main of o2-mft-reco-workflow, device shutting down. Reason: Fatal error
./sim-challenge.log[ERROR] Workflow crashed - PID 14846 (internal-dpl-ccdb-backend) did not exit correctly however it's not clear why. Exit code forced to 128.
./sim-challenge.log[ERROR]  - Device internal-dpl-ccdb-backend: pid 14846 (exit 128)
./sim-challenge.log[INFO]    - First error: [12:19:04][FATAL] Unable to find object CTP/Calib/OrbitReset/1546300800000
./sim-challenge.log[ERROR] SEVERE: Device internal-dpl-ccdb-backend (14846) had at least one message above severity 5: Unable to find object CTP/Calib/OrbitReset/1546300800000
./digi.log[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/37}
./digi.log[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/38}
./digi.log[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/40}
./digi.log[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/42}
./digi.log[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/43}
./digi.log[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/45}
./digi.log[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/46}
./digi.log[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/47}
./digi.log[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/48}
./digi.log[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/49}
./digi.log[8756:TRDDigitizer]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/37}
./digi.log[8756:TRDDigitizer]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/38}
./digi.log[8756:TRDDigitizer]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/40}
./digi.log[8756:TRDDigitizer]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/42}
./digi.log[8756:TRDDigitizer]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/43}
./digi.log[8756:TRDDigitizer]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/45}
./digi.log[8756:TRDDigitizer]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/46}
./digi.log[8756:TRDDigitizer]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/47}
./digi.log[8756:TRDDigitizer]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/48}
./digi.log[8756:TRDDigitizer]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/49}
./digi.log[8762:TPCDigitizer_4]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/37}
[0 more errors; see full log]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/13801/8222c00d531b3b169aea53225e397cc735660086/build_O2_fullCI/pretty.html).

--- a/Generators/include/Generators/BoxGenerator.h
+++ b/Generators/include/Generators/BoxGenerator.h
@@ -18,6 +18,7 @@
 #include "TParticle.h"
 #include <vector>
 #include <Generators/BoxGunParam.h>
+#include "SimulationDataFormat/MCEventHeader.h"
 
 namespace o2::eventgen
 {
@@ -90,6 +91,14 @@ class BoxGenerator : public Generator
     mParticles.clear();
     std::copy(mEvent.begin(), mEvent.end(), std::back_insert_iterator(mParticles));
     return true;
+  }
+
+  void updateHeader(o2::dataformats::MCEventHeader* eventHeader) override
+  {
+    using Key = o2::dataformats::MCInfoKeys;
+    if (eventHeader) {
+      eventHeader->putInfo<std::string>(Key::generator, "o2::eventgen::BoxGenerator");
+    }
   }
 
  private:

--- a/Generators/include/Generators/Generator.h
+++ b/Generators/include/Generators/Generator.h
@@ -73,6 +73,7 @@ class Generator : public FairGenerator
   /** methods to override **/
   virtual Bool_t generateEvent() = 0;   // generates event (in structure internal to generator)
   virtual Bool_t importParticles() = 0; // fills the mParticles vector (transfer from generator state)
+  virtual void updateHeader(o2::dataformats::MCEventHeader* eventHeader) {};
 
   /** setters **/
   void setMomentumUnit(double val) { mMomentumUnit = val; };
@@ -101,9 +102,6 @@ class Generator : public FairGenerator
   Generator(const Generator&);
   /** operator= **/
   Generator& operator=(const Generator&);
-
-  /** methods that can be overridded **/
-  virtual void updateHeader(o2::dataformats::MCEventHeader* eventHeader){};
 
   /** internal methods **/
   Bool_t addTracks(FairPrimaryGenerator* primGen);

--- a/Generators/include/Generators/GeneratorFromFile.h
+++ b/Generators/include/Generators/GeneratorFromFile.h
@@ -135,9 +135,11 @@ class GeneratorFromEventPool : public o2::eventgen::Generator
   }
   bool importParticles() override
   {
+    mO2KineGenerator->clearParticles(); // clear old container before filling with new ones
     auto import_good = mO2KineGenerator->importParticles();
     // transfer the particles (could be avoided)
     mParticles = mO2KineGenerator->getParticles();
+
     return import_good;
   }
 

--- a/Generators/include/Generators/GeneratorHybrid.h
+++ b/Generators/include/Generators/GeneratorHybrid.h
@@ -61,6 +61,7 @@ class GeneratorHybrid : public Generator
   Bool_t Init() override;
   Bool_t generateEvent() override;
   Bool_t importParticles() override;
+  void updateHeader(o2::dataformats::MCEventHeader* eventHeader) override;
 
   void setNEvents(int n) { mNEvents = n; }
 
@@ -106,6 +107,7 @@ class GeneratorHybrid : public Generator
   bool mIsInitialized = false;
 
   int mNEvents = -1; // the number of events to be done, if known (helps initiating cleanup)
+  o2::dataformats::MCEventHeader mMCEventHeader; // to capture event headers
 
   enum class GenMode {
     kSeq,

--- a/Generators/src/GeneratorFromFile.cxx
+++ b/Generators/src/GeneratorFromFile.cxx
@@ -398,10 +398,12 @@ bool GeneratorFromEventPool::Init()
     LOG(error) << "No file found that can be used with EventPool generator";
     return false;
   }
+  LOG(info) << "Found " << mPoolFilesAvailable.size() << " available event pool files";
 
   // now choose the actual file
-  std::uniform_int_distribution<int> distribution(0, mPoolFilesAvailable.size());
-  mFileChosen = mPoolFilesAvailable[distribution(mRandomEngine)];
+  std::uniform_int_distribution<int> distribution(0, mPoolFilesAvailable.size() - 1);
+  auto chosenIndex = distribution(mRandomEngine);
+  mFileChosen = mPoolFilesAvailable[chosenIndex];
   LOG(info) << "EventPool is using file " << mFileChosen;
 
   // we bring up the internal mO2KineGenerator

--- a/Generators/src/GeneratorHybrid.cxx
+++ b/Generators/src/GeneratorHybrid.cxx
@@ -332,17 +332,15 @@ bool GeneratorHybrid::importParticles()
   LOG(info) << "Importing particles for task " << genIndex;
 
   // at this moment the mIndex-th generator is ready to be used
-  std::copy(gens[genIndex]->getParticles().begin(), gens[genIndex]->getParticles().end(), std::back_insert_iterator(mParticles));
+  mParticles.clear();
+  mParticles = gens[genIndex]->getParticles();
+
+  // fetch the event Header information from the underlying generator
+  mMCEventHeader.clearInfo();
+  gens[genIndex]->updateHeader(&mMCEventHeader);
 
   mInputTaskQueue.push(genIndex);
   mTasksStarted++;
-
-  // we need to fix particles statuses --> need to enforce this on the importParticles level of individual generators
-  for (auto& p : mParticles) {
-    auto st = o2::mcgenstatus::MCGenStatusEncoding(p.GetStatusCode(), p.GetStatusCode()).fullEncoding;
-    p.SetStatusCode(st);
-    p.SetBit(ParticleStatus::kToBeDone, true);
-  }
 
   mseqCounter++;
   mEventCounter++;
@@ -351,6 +349,17 @@ bool GeneratorHybrid::importParticles()
     mStopFlag = true;
   }
   return true;
+}
+
+void GeneratorHybrid::updateHeader(o2::dataformats::MCEventHeader* eventHeader)
+{
+  if (eventHeader) {
+    // we forward the original header information if any
+    eventHeader->copyInfoFrom(mMCEventHeader);
+
+    // put additional information about
+    eventHeader->putInfo<std::string>("forwarding-generator", "HybridGen");
+  }
 }
 
 template <typename T>

